File: packages/suspension_multibody/src/suspension_multibody/cases/ride_four_post.py
"""
Emit and run the `ride_four_post` contract documents.

A four-post rig prescribes the height of each wheel pad.  The family exists so
that the excitation is written down once, declaratively -- per corner offset,
amplitude, frequency and phase -- and the kernel expands it, rather than every
caller hand-sampling sinusoids and hoping to agree on the derivative.

The expansion is a pure function of the declaration, so it is checked the only
way such a thing can be: the same excitation is computed independently here and
handed to the *vehicle* family as an explicit table, and the two runs have to
agree.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

import numpy as np

from ..axle_dynamics.schema import AxleSolverSettings
from .vehicle_dynamic import _solver_block

__all__ = ["FourPostCorner", "case_document"]


@dataclass(frozen=True)
class FourPostCorner:
    """One pad: which wheel it carries and the shape it is driven with."""

    tire: str
    offset_m: float = 0.0
    amplitude_m: float = 0.0
    frequency_hz: float = 0.0
    phase_rad: float = 0.0


def case_document(
    *,
    name: str,
    corners: tuple[FourPostCorner, ...],
    times_s: tuple[float, ...],
    settings: AxleSolverSettings,
) -> dict[str, Any]:
    """Describe a four-post sweep as a case document."""
    times = np.asarray(times_s, dtype=float)
    if times.size < 2:
        raise ValueError("a four-post case needs at least two sample times")
    steps = np.diff(times)
    if not np.allclose(steps, steps[0], rtol=0.0, atol=1e-15):
        raise ValueError(
            "the contract time grid is a start/end/step, so the samples must be uniform"
        )
    if not corners:
        raise ValueError("a four-post case needs at least one corner")
    return {
        "contract": "multibody-case",
        "contract_version": 1,
        "kind": "case",
        "family": "ride_four_post",
        "name": name,
        "time": {
            "start_s": float(times[0]),
            "end_s": float(times[-1]),
            "step_s": float(steps[0]),
        },
        "solver": _solver_block(settings),
        "four_post": {
            "corners": [
                {
                    "tire": corner.tire,
                    "offset_m": float(corner.offset_m),
                    "amplitude_m": float(corner.amplitude_m),
                    "frequency_hz": float(corner.frequency_hz),
                    "phase_rad": float(corner.phase_rad),
                }
                for corner in corners
            ]
        },
    }


def corner_signals(
    corners: tuple[FourPostCorner, ...], times_s: tuple[float, ...]
) -> tuple[dict[str, np.ndarray], dict[str, np.ndarray]]:
    """
    Return the height and velocity each corner is driven with.

    This is the same expansion the kernel performs, written independently, and
    it is what the equivalence test uses as the other side of the comparison.
    The velocity is the analytic derivative: a finite difference of the height
    is a different excitation.
    """
    times = np.asarray(times_s, dtype=float)
    height: dict[str, np.ndarray] = {}
    velocity: dict[str, np.ndarray] = {}
    for corner in corners:
        angular = 2.0 * np.pi * float(corner.frequency_hz)
        angle = angular * times + float(corner.phase_rad)
        height[corner.tire] = float(corner.offset_m) + float(corner.amplitude_m) * np.sin(angle)
        velocity[corner.tire] = float(corner.amplitude_m) * angular * np.cos(angle)
    return height, velocity

File: packages/suspension_kernel/cpp/src/cases/ride_four_post.cpp
/// Expand a `ride_four_post` case document into one solver run.
///
/// A four-post rig prescribes the height of each wheel pad, so this family's
/// whole job is turning a declarative excitation -- per-corner offset,
/// amplitude, frequency and phase -- into the per-tire road height and road
/// velocity the kernel interpolates.  Nothing else about the run differs from a
/// vehicle case: the model decides what the excitation does to the car.
///
/// The velocity is the analytic derivative rather than a difference quotient.
/// A sampled height whose velocity is a finite difference of itself is a
/// slightly different excitation from the one that was asked for, and the
/// kernel interpolates both independently.
///
/// The document may also carry a measured height table per corner instead of a
/// sinusoid; the two forms are exclusive and a corner that declares neither is
/// an error, because a silently flat pad looks exactly like a rig that is
/// working and measuring nothing.

#include "case_common.hpp"

#include "mb_config/constants.hpp"

#include <cmath>
#include <string>
#include <vector>

namespace axle_kernel {
namespace case_detail {
namespace {

using Json = JsonValue;

struct Corner {
  int tire = -1;
  double offset = 0.0;
  double amplitude = 0.0;
  double frequency = 0.0;
  double phase = 0.0;
  /// A measured pad history, when the corner carries one instead of a shape.
  std::string height_blob;
  std::string velocity_blob;
};

}  // namespace

bool expand_ride_four_post(const Json& document, const std::string& blob,
                           const ContractModel& model, ContractPlan& out,
                           std::string& error) {
  if (!read_identity(document, "ride_four_post", out, error)) return false;

  std::vector<double> times;
  if (!read_time(document, blob, times, error)) return false;
  const std::size_t sample_count = times.size();
  const std::size_t tire_count = model.tire_order().size();

  const Json* four_post = document.find("four_post");
  if (four_post == nullptr || !four_post->is_object()) {
    return fail(error, "a ride_four_post case needs a four_post section");
  }
  const Json* corners = four_post->find("corners");
  if (corners == nullptr || !corners->is_array() || corners->items.empty()) {
    return fail(error, "four_post.corners must be a non-empty array");
  }

  std::vector<Corner> plan;
  for (const Json& corner : corners->items) {
    if (!corner.is_object()) {
      return fail(error, "a four_post corner is not an object");
    }
    const std::string* tire = corner.find_string("tire");
    if (tire == nullptr) return fail(error, "a four_post corner needs a tire name");
    const int index = model.tire_index(*tire);
    if (index < 0) {
      return fail(error, "four_post names unknown tire " + *tire);
    }
    Corner entry;
    entry.tire = index;
    entry.offset = number_or(corner, "offset_m", 0.0);
    entry.amplitude = number_or(corner, "amplitude_m", 0.0);
    entry.frequency = number_or(corner, "frequency_hz", 0.0);
    entry.phase = number_or(corner, "phase_rad", 0.0);
    if (entry.amplitude < 0.0 || entry.frequency < 0.0) {
      return fail(error, "four_post corner " + *tire +
                              " has a negative amplitude or frequency");
    }
    if (const std::string* name = corner.find_string("blob"); name != nullptr) {
      entry.height_blob = *name;
    }
    if (const std::string* name = corner.find_string("velocity_blob");
        name != nullptr) {
      entry.velocity_blob = *name;
    }
    plan.push_back(entry);
  }

  ContractCase run;
  run.name = out.name;
  run.sample_count = sample_count;
  run.sample_times = times;
  run.road_z.assign(sample_count * tire_count, 0.0);
  run.road_velocity.assign(sample_count * tire_count, 0.0);
  run.body_wrench.assign(sample_count * model.body_count() * 6, 0.0);
  run.wheel_torque.assign(sample_count * tire_count, 0.0);
  run.brake_torque.assign(sample_count * tire_count, 0.0);
  run.steering_target.assign(sample_count * model.steering_order().size(), 0.0);
  run.steering_rate.assign(sample_count * model.steering_order().size(), 0.0);
  if (!read_road(document, run, error)) return false;
  if (!read_case_overrides(document, run, error)) return false;
  if (run.has_static_gauge) {
    const int index = model.body_index(run.static_gauge_body_name);
    if (index < 0) {
      return fail(error, "static_gauge names unknown body " + run.static_gauge_body_name);
    }
    run.static_gauge_body = static_cast<std::size_t>(index);
  }

  for (const Corner& corner : plan) {
    const std::size_t column = static_cast<std::size_t>(corner.tire);
    const double angular = 2.0 * kPi * corner.frequency;
    // A measured history is read through the same described-array path every
    // other payload table uses, so a truncated one is an error.
    std::vector<double> measured;
    std::vector<double> measured_velocity;
    if (!corner.height_blob.empty()) {
      if (!read_described_array(document, blob, corner.height_blob, sample_count,
                                measured, error)) {
        return false;
      }
      if (!corner.velocity_blob.empty() &&
          !read_described_array(document, blob, corner.velocity_blob, sample_count,
                                measured_velocity, error)) {
        return false;
      }
    }
    for (std::size_t sample = 0; sample < sample_count; ++sample) {
      const std::size_t slot = sample * tire_count + column;
      if (!corner.height_blob.empty()) {
        run.road_z[slot] = measured[sample];
        // A measured height with no measured velocity falls back to the
        // analytic derivative of the shape it was declared with, which is
        // exactly zero when the corner declared no shape.
        run.road_velocity[slot] = measured_velocity.empty()
                                      ? corner.amplitude * angular *
                                            std::cos(angular * times[sample] + corner.phase)
                                      : measured_velocity[sample];
        continue;
      }
      const double angle = angular * times[sample] + corner.phase;
      run.road_z[slot] = corner.offset + corner.amplitude * std::sin(angle);
      run.road_velocity[slot] = corner.amplitude * angular * std::cos(angle);
    }
  }

  out.cases.push_back(std::move(run));
  return true;
}

}  // namespace case_detail
}  // namespace axle_kernel
